src/modules/rlm_smtp/rlm_smtp.c

```diff
@@ -0,0 +1,614 @@
+	vp = fr_cursor_current(&mail_ctx->body_cursor);
+
+	/* Copy the vp into the email. If it cannot all be loaded, return the amount of memory that was loaded and get called again */
+	if (fr_dbuff_memcpy_in_partial(&out, &mail_ctx->vp_in, SIZE_MAX) < fr_dbuff_remaining(&mail_ctx->vp_in)) {
+		RDEBUG2("%zu bytes used (partial copy)", fr_dbuff_used(&out));
+		return fr_dbuff_used(&out);
+	}
+	/* Once this value pair is fully copied, prepare for the next element */
+	vp = fr_cursor_next(&mail_ctx->body_cursor);
+	if (vp) {
+		fr_dbuff_init(&mail_ctx->vp_in, (uint8_t const *)vp->vp_strvalue, vp->vp_length);
+
+	}
+	RDEBUG2("%zu bytes used (full copy)", fr_dbuff_used(&out));
+	return fr_dbuff_used(&out);
+}
+
+/*
+ * Adds every SMTP_Attachments file to the email as a MIME part
+ */
+static int attachments_source(fr_mail_ctx *uctx, curl_mime *mime, rlm_smtp_t const *inst)
+{
+	curl_mimepart		*part;
+	REQUEST			*request = uctx->request;
+	VALUE_PAIR 		*vp;
+	int 			attachments_set = 0;
+
+	vp = fr_cursor_iter_by_da_init(&uctx->cursor, &uctx->request->packet->vps, attr_smtp_attachment_file);
+
+	/* Initialize the buffer to write the file path */
+	fr_sbuff_init_talloc(uctx, &uctx->path_buffer, &uctx->sbuff_ctx, 0, SIZE_MAX);
+
+	/* Write the initial path to the buffer */
+	fr_sbuff_in_bstrncpy(&uctx->path_buffer, inst->template_dir, strlen(inst->template_dir));
+	/* Make sure the template_directory path ends in a "/" */
+	if (inst->template_dir[strlen(inst->template_dir)-1] != '/'){
+		RDEBUG2("Adding / to end of template_dir");
+		fr_sbuff_in_bstrncpy(&uctx->path_buffer, "/", 1);
+	}
+	/* Mark the buffer so we only re-write after the template_dir component */
+	fr_sbuff_marker(&uctx->m, &uctx->path_buffer);
+
+	/* Check for any file attachments */
+	while(vp) {
+		/* Move to the end of the template directory filepath */
+		fr_sbuff_set(&uctx->path_buffer, &uctx->m);
+
+		/* Check to see if the file attachment is valid, skip it if not */
+		RDEBUG2("Trying to set attachment: %s", vp->vp_strvalue);
+		if(vp->vp_tainted) {
+			RDEBUG2("Skipping a tainted attachment");
+			goto next;
+		}
+		if(strncmp(vp->vp_strvalue, "/", 1) == 0) {
+			RDEBUG2("File attachments cannot be an absolute path");
+			goto next;
+		}
+		if(strncmp(vp->vp_strvalue, "..", 2) == 0) {
+			RDEBUG2("Cannot access values outside of template_directory");
+			goto next;
+		}
+
+		/* Copy the filename into the buffer */
+		fr_sbuff_in_bstrncpy(&uctx->path_buffer, vp->vp_strvalue, vp->vp_length);
+		fr_sbuff_trim_talloc(&uctx->path_buffer, SIZE_MAX);
+
+		/* Add the file attachment as a mime encoded part */
+		attachments_set++;
+		RDEBUG2("ititializing attachment: %s", uctx->path_buffer.buff);
+		part = curl_mime_addpart(mime);
+		curl_mime_encoder(part, "base64");
+		curl_mime_filedata(part, uctx->path_buffer.buff);
+	next:
+		vp = fr_cursor_next(&uctx->cursor);
+	}
+	RDEBUG2("Ititialized %d attachment(s)", attachments_set);
```

Comment:
`talloc_free(uctx->path_buffer.buff)`

Reply:
Everything is freed in it's appropriate function. No memory leaks on travis